### ISSUE TITLE
Treesitter's `descendant_for_range` documentation doesn't clarify 0- or 1-based indexing for columns

### ISSUE THREAD

**Issue by lucario387** [neovim/neovim#26569]
### Problem

Doing the test with a json file, I find the returned result of `TSNode:descendant_for_range` doesn't actually reflect the node to cover for that range, or the documentation isn't clear enough about columns indexing.

### Steps to reproduce

- minimal.lua
```lua
-- Run this file as `nvim --clean -u minimal.lua`

for name, url in pairs{
  -- ADD PLUGINS _NECESSARY_ TO REPRODUCE THE ISSUE, e.g:
  ["nvim-treesitter"] = 'https://github.com/nvim-treesitter/nvim-treesitter'
} do
  local install_path = vim.fn.fnamemodify('nvim_issue/'..name, ':p')
  if vim.fn.isdirectory(install_path) == 0 then
    vim.fn.system { 'git', 'clone', '--depth=1', url, install_path }
  end
  vim.opt.runtimepath:append(install_path)
end

require("nvim-treesitter.configs").setup({
  ensure_installed = { "json" },
})

vim.keymap.set("n", "qq", function()
  local lnum = 2
  local parser = vim.F.npcall(vim.treesitter.get_parser, 0)
  local root = parser:parse(true)[1]:root()
  local line = vim.api.nvim_buf_get_lines(0, lnum, lnum + 1, false)[1] or ""
  local scol = vim.fn.indent(lnum + 1)
  local ecol = #line - 1
  vim.print(root:descendant_for_range(lnum, scol, lnum, ecol):type())
end)
```
- repro.json
```json
{
  "foo": [
    "bar",
    "baz",
  ]
}
```

- `nvim --clean -u minimal.lua repro.json`
- While in `repro.json`, press `qq`

### Expected behavior

I expected the node type returned to be `array`, and not `string`, as `string` node doesn't cover the final `,` at end of line 3

### Neovim version (nvim -v)

v0.10.0-dev-36552adb3

### Vim (not Nvim) behaves the same?

N/A

### Operating system/version

Linux 6.6.6

### Terminal name/version

kitty 0.31.0

### $TERM environment variable

xterm-kitty

### Installation

Build from source

**Comment by lucario387** [comment]
Just edited the issue text. I also found the behavior for `descendant_for_range` also seems to be inconsistent.

**Comment by wookayin** [comment]
In this example, the exact argument it's being called with: `descendant_for_range(2, 4, 2, 9) == "string"`.

```
  012345678
0 {
1   "foo": [
2     "bar",
3     "baz",
4   ]
5 }
```

so this Range4 (0-indexed, end-exclusive) represents `"bar"`. `row = 2`, col = `[4, 9)` (in vim, Line 3, and . So I guess "string" is correct?

**Comment by lucario387** [comment]
Not really, if `descendant_for_range(2, 4, 2, 9)` returns `string`, then shouldn't the range `2, 9, 2, 9`, which is part of `2, 4, 2, 9`, return the same value as before? But instead, it returns `,`. As the 2 nodes have a common parent of `array` which has both nodes, shouldn't the smallest node actually spanning the range `(2, 4, 2, 9)` be `array` instead of `string`

**Comment by lucario387** [comment]
I don't know if the intended behavior of `descendant_for_range` to be end-exclusive, or end-inclusive, but having some stuffs to clear the misunderstandings will be nice.

**Comment by wookayin** [comment]
Range4(2, 9, 2, 9) represents *an empty range* (starting at `,` right after `"bar"`) and I guess the behavior on length-0 range is somewhat undefined. It feels like it were treated as if a range `(2, 9, 2, 10)` with length 1. So `,` is correct or at least not surprising.

I think Range(2, 4, 2, 9) is **not** included in Range(2, 9, 2, 9), because Range(2, 4, 2, 9) *excludes* col = 9 (0-indexed). Past-to-end indexing is confusing.

**Comment by lucario387** [comment]
I do understand that, guess it's just not documented. Also means nvim-treesitter has been incorrectly (yet correctly) using `descendant_for_range` with length-0 range
